DashboardClient/DashboardClient.cpp

```diff
@@ -27,6 +27,173 @@ namespace Umati
+						for(std::list<ModelOpcUa::PlaceholderElement>::iterator it = instances.begin(); it != instances.end(); it++) {
+							deleteAndUnsubscribeNode(*it);
+						}
+
+					}
+				}
+			}
+			LOG(INFO) << "Erase NodeId :" <<  element->NodeId;
+			browsedNodes.erase(element->NodeId);
+			browsedSimpleNodes.erase(element->NodeId);
+
+		}
+		void DashboardClient::deleteAndUnsubscribeNode(const ModelOpcUa::SimpleNode simpleNode) {
+			const std::list<std::shared_ptr<const ModelOpcUa::Node>> children = simpleNode.ChildNodes;
+			for(auto it = children.begin(); it != children.end();it++) {
+				std::shared_ptr<const ModelOpcUa::Node> node = *it;
+				std::shared_ptr<const ModelOpcUa::SimpleNode> simpleNode = std::dynamic_pointer_cast<const ModelOpcUa::SimpleNode>(node);
+				if(simpleNode != nullptr) {
+					const ModelOpcUa::SimpleNode simpleN = *simpleNode;
+					deleteAndUnsubscribeNode(simpleN);
+				} else {
+					std::shared_ptr<const ModelOpcUa::PlaceholderNode> placeHolderNode = std::dynamic_pointer_cast<const ModelOpcUa::PlaceholderNode>(node);
+					if(placeHolderNode != nullptr) {
+						std::list<ModelOpcUa::PlaceholderElement> instances = placeHolderNode->getInstances();
+						for(std::list<ModelOpcUa::PlaceholderElement>::iterator it = instances.begin(); it != instances.end(); it++) {
+							deleteAndUnsubscribeNode(*it);
+						}
+
+					}
+				}
+			}
+			LOG(INFO) << "Erase NodeId :" << simpleNode.NodeId;
+			browsedNodes.erase(simpleNode.NodeId);
+			browsedSimpleNodes.erase(simpleNode.NodeId);
+		}
+		void DashboardClient::subscribeEvents() {
+			auto ecbf = [](IDashboardDataClient::StructureChangeEvent sce, void* context) {
+				DashboardClient* pDashboardClient = static_cast<DashboardClient*> (context);
+				if(pDashboardClient != nullptr) {
+					LOG(INFO) << "Event Callback" << pDashboardClient;
+					if(sce.nodeAdded || sce.referenceAdded) {
+							pDashboardClient->updateAddDataSet(sce.refreshNode);
+							pDashboardClient->Publish();
+						}
+						if(sce.nodeDeleted || sce.referenceDeleted) {
+							pDashboardClient->updateDeleteDataSet(sce.refreshNode);
+							pDashboardClient->Publish();
+						}
+				}
+			};
+			m_pDashboardDataClient->SubscribeEvent(ecbf, this);
+		}
+		void DashboardClient::updateAddDataSet(ModelOpcUa::NodeId_t refreshNodeId) {
+			LOG(INFO) << "Update Add DataSet";
+			if(!m_dataSets.empty()) {
+				std::shared_ptr<Umati::Dashboard::DashboardClient::DataSetStorage_t> dataSet = m_dataSets.front();
+				auto search = browsedSimpleNodes.find(refreshNodeId);
+				if( search != browsedSimpleNodes.end()) {
+					std::shared_ptr<const ModelOpcUa::SimpleNode> simpleNode = search->second;
+					auto childNodes = simpleNode->ChildNodes;
+					if(!childNodes.empty()) {
+						auto child = childNodes.front();
+						std::shared_ptr<const ModelOpcUa::PlaceholderNode> placeholderNode = std::dynamic_pointer_cast<const ModelOpcUa::PlaceholderNode>(child);
+						if(placeholderNode == nullptr) {
+						} else {
```

Comment:
Use `return` and write the else-Code without indentation.

Reply:
Done.